File: backend/smartapi.js
/**
 * SmartAPI Client — Angel One SmartAPI helper
 * Handles authentication, TOTP generation, and API calls
 */

const axios = require('axios');

// Try to load totp-generator safely
let totp;
try {
    totp = require('totp-generator');
    console.log('[SmartAPI] totp-generator loaded ✓');
} catch (e) {
    console.warn('[SmartAPI] totp-generator NOT found — TOTP will be 000000 (login will fail if TOTP is required)');
    totp = null;
}

const BASE_URL = 'https://apiconnect.angelone.in';

class SmartAPIClient {
    constructor(config) {
        this.apiKey      = config.apiKey;
        this.clientId    = config.clientId;
        this.password    = config.password;
        this.totpSecret  = config.totpSecret;

        this.jwtToken    = null;
        this.feedToken   = null;
        this.refreshToken = null;
        this.tokenExpiry = null;
        this.isDummy     = this._isDummyConfig(config);

        if (this.isDummy) {
            console.warn('[SmartAPI] ⚠️  Credentials incomplete — will run in MOCK mode:');
            if (!config.apiKey)               console.warn('           • ANGEL_API_KEY is missing');
            if (!config.clientId)             console.warn('           • ANGEL_CLIENT_ID is missing');
            if (!config.password)             console.warn('           • ANGEL_PASSWORD is missing');
        } else {
            console.log(`[SmartAPI] ✅ Credentials loaded for client: ${this.clientId}`);
        }
    }

    _isDummyConfig(c) {
        return !c.apiKey     || c.apiKey.startsWith('DUMMY')     ||
               !c.clientId  || c.clientId.startsWith('DUMMY')   ||
               !c.password  || c.password.startsWith('DUMMY');
    }

    _generateTOTP() {
        if (!totp || !this.totpSecret || this.totpSecret.startsWith('DUMMY')) {
            console.warn('[SmartAPI] Generating dummy TOTP 000000 — real TOTP not available');
            return '000000';
        }
        try {
            // totp-generator v0.0.14 API: totp(secret) returns string
            // v1.x API: totp.generate(secret) returns { otp, expires }
            const result = typeof totp === 'function'
                ? totp(this.totpSecret)
                : totp.generate(this.totpSecret);
            const code = typeof result === 'object' ? result.otp : result;
            console.log(`[SmartAPI] TOTP generated: ${code}`);
            return code;
        } catch (err) {
            console.error('[SmartAPI] TOTP generation error:', err.message);
            return '000000';
        }
    }

    _headers(withAuth = false) {
        const h = {
            'Content-Type':        'application/json',
            'Accept':              'application/json',
            'X-UserType':          'USER',
            'X-SourceID':          'WEB',
            'X-ClientLocalIP':     '127.0.0.1',
            'X-ClientPublicIP':    '127.0.0.1',
            'X-MACAddress':        '00:00:00:00:00:00',
            'X-PrivateKey':        this.apiKey,
        };
        if (withAuth && this.jwtToken) {
            h['Authorization'] = `Bearer ${this.jwtToken}`;
        }
        return h;
    }

    isTokenValid() {
        return !!(this.jwtToken && this.tokenExpiry && Date.now() < this.tokenExpiry);
    }

    async login() {
        if (this.isDummy) {
            console.log('[SmartAPI] Dummy credentials — skipping login, staying in MOCK mode');
            return { success: false, reason: 'dummy_credentials' };
        }

        try {
            const totpCode = this._generateTOTP();
            console.log(`[SmartAPI] Attempting login for client: ${this.clientId} with TOTP: ${totpCode}`);

            const res = await axios.post(
                `${BASE_URL}/rest/auth/angelbroking/user/v1/loginByPassword`,
                {
                    clientcode: this.clientId,
                    password:   this.password,
                    totp:       totpCode,
                },
                { headers: this._headers(), timeout: 15000 }
            );

            console.log('[SmartAPI] Login response status:', res.data?.status);
            console.log('[SmartAPI] Login response message:', res.data?.message);

            if (res.data?.status && res.data?.data?.jwtToken) {
                this.jwtToken     = res.data.data.jwtToken;
                this.feedToken    = res.data.data.feedToken;
                this.refreshToken = res.data.data.refreshToken;
                // Tokens expire at midnight IST — cache for 6 hours
                this.tokenExpiry  = Date.now() + (6 * 60 * 60 * 1000);
                console.log('[SmartAPI] ✅ Login successful! LIVE MODE active.');
                return { success: true };
            }

            const msg = res.data?.message || JSON.stringify(res.data) || 'Login failed — no JWT in response';
            console.error('[SmartAPI] ❌ Login failed:', msg);
            return { success: false, reason: msg };

        } catch (err) {
            const msg = err.response?.data?.message || err.message;
            const status = err.response?.status;
            console.error(`[SmartAPI] ❌ Login HTTP error (${status}):`, msg);
            if (err.response?.data) {
                console.error('[SmartAPI] Full error response:', JSON.stringify(err.response.data));
            }
            return { success: false, reason: msg };
        }
    }

    async ensureAuthenticated() {
        if (!this.isTokenValid()) {
            return await this.login();
        }
        return { success: true };
    }

    /**
     * Get market quotes for a list of instruments
     * @param {Array} instruments - [{ exchange: 'NSE', tradingsymbol: 'NIFTY', symboltoken: '26000' }]
     * @param {string} mode - 'LTP' | 'OHLC' | 'FULL'
     */
    async getQuote(instruments, mode = 'FULL') {
        const authResult = await this.ensureAuthenticated();
        if (!authResult.success) return null;

        try {
            const res = await axios.post(
                `${BASE_URL}/rest/secure/angelbroking/market/v1/quote/`,
                { mode, exchangeTokens: this._groupByExchange(instruments) },
                { headers: this._headers(true), timeout: 8000 }
            );
            return res.data?.data || null;
        } catch (err) {
            console.error('[SmartAPI] Quote error:', err.response?.data?.message || err.message);
            return null;
        }
    }

    async getLtpData(exchange, tradingsymbol, symboltoken) {
        const authResult = await this.ensureAuthenticated();
        if (!authResult.success) return null;

        try {
            const res = await axios.post(
                `${BASE_URL}/rest/secure/angelbroking/order/v1/getLtpData`,
                { exchange, tradingsymbol, symboltoken },
                { headers: this._headers(true), timeout: 5000 }
            );
            return res.data?.data || null;
        } catch (err) {
            console.error('[SmartAPI] LTP error:', err.response?.data?.message || err.message);
            return null;
        }
    }

    /**
     * Get option chain for a symbol
     */
    async getOptionChain(symbol, expiryDate, strikePrice) {
        const authResult = await this.ensureAuthenticated();
        if (!authResult.success) return null;

        try {
            const res = await axios.post(
                `${BASE_URL}/rest/secure/angelbroking/derivatives/v1/optionchain`,
                { name: symbol, expirydate: expiryDate, strikePrice, optionType: 'CE PE' },
                { headers: this._headers(true), timeout: 10000 }
            );
            return res.data?.data || null;
        } catch (err) {
            console.error('[SmartAPI] OptionChain error:', err.response?.data?.message || err.message);
            return null;
        }
    }

    _groupByExchange(instruments) {
        const grouped = {};
        instruments.forEach(({ exchange, symboltoken }) => {
            if (!grouped[exchange]) grouped[exchange] = [];
            grouped[exchange].push(symboltoken);
        });
        return grouped;
    }
}

module.exports = SmartAPIClient;

File: backend/app/smartapi.py
"""
smartapi.py — Async Angel One SmartAPI client for AntiGravity backend.
"""

from __future__ import annotations

import asyncio
import logging
import time
from typing import Any, Optional

import httpx
import pyotp

from app.config import Settings

logger = logging.getLogger(__name__)

# ------------------------------------------------------------------
# Constants
# ------------------------------------------------------------------

_ANGEL_BASE = "https://apiconnect.angelone.in"

_HISTORICAL_API = (
    f"{_ANGEL_BASE}/rest/secure/angelbroking/historical/v1/getCandleData"
)

_TOKEN_TTL_SEC = 6 * 60 * 60

_login_lock = asyncio.Lock()


# ------------------------------------------------------------------
# SmartAPI Client
# ------------------------------------------------------------------

class SmartAPIClient:
    """Async Angel One SmartAPI wrapper."""

    def __init__(self, settings: Settings) -> None:

        self._api_key = settings.angel_api_key
        self._client_id = settings.angel_client_id
        self._password = settings.angel_password
        self._totp_secret = settings.angel_totp_secret

        self.jwt_token: Optional[str] = None
        self.feed_token: Optional[str] = None
        self.refresh_token: Optional[str] = None

        self.token_expiry = 0.0

        logger.info(
            "[SmartAPI] Client initialised : %s",
            self._client_id,
        )
    # --------------------------------------------------------------

    def _generate_totp(self) -> str:

        try:

            return pyotp.TOTP(
                self._totp_secret
            ).now()

        except Exception as e:

            logger.exception(e)

            return "000000"

    # --------------------------------------------------------------

    def _base_headers(
        self,
        *,
        with_auth: bool = False,
    ) -> dict[str, str]:

        headers = {

            "Content-Type": "application/json",

            "Accept": "application/json",

            "X-UserType": "USER",

            "X-SourceID": "WEB",

            "X-ClientLocalIP": "127.0.0.1",

            "X-ClientPublicIP": "127.0.0.1",

            "X-MACAddress": "00:00:00:00:00:00",

            "X-PrivateKey": self._api_key,

        }

        if with_auth and self.jwt_token:

            headers["Authorization"] = (
                f"Bearer {self.jwt_token}"
            )

        return headers

    # --------------------------------------------------------------

    def _group_by_exchange(
        self,
        instruments: list[dict[str, str]],
    ) -> dict[str, list[str]]:

        grouped: dict[str, list[str]] = {}

        for instrument in instruments:

            exchange = instrument.get(
                "exchange",
                "NSE",
            )

            token = instrument.get(
                "symboltoken",
                "",
            )

            grouped.setdefault(
                exchange,
                [],
            ).append(token)

        return grouped

    # --------------------------------------------------------------

    def is_token_valid(self) -> bool:

        return (

            self.jwt_token is not None

            and

            time.time() < self.token_expiry

        )

    # --------------------------------------------------------------

    async def login(self) -> dict[str, Any]:

        async with _login_lock:

            if self.is_token_valid():

                return {
                    "success": True
                }

            totp = self._generate_totp()

            logger.info(
                "[SmartAPI] Login..."
            )

            try:

                async with httpx.AsyncClient(
                    timeout=15
                ) as client:

                    response = await client.post(

                        f"{_ANGEL_BASE}/rest/auth/angelbroking/user/v1/loginByPassword",

                        json={

                            "clientcode": self._client_id,

                            "password": self._password,

                            "totp": totp,

                        },

                        headers=self._base_headers(),

                    )

                if response.status_code != 200:
                    logger.error(
                        "[Historical] %s\n%s",
                        response.status_code,
                        response.text,
                    )
                    return []

                try:
                    data = response.json()
                except Exception:
                    logger.error(response.text)
                    return []

                if (

                    data.get("status")

                    and

                    data.get("data", {}).get("jwtToken")

                ):

                    auth = data["data"]

                    self.jwt_token = auth["jwtToken"]

                    self.feed_token = auth.get(
                        "feedToken"
                    )

                    self.refresh_token = auth.get(
                        "refreshToken"
                    )

                    self.token_expiry = (
                        time.time()
                        + _TOKEN_TTL_SEC
                    )

                    logger.info(
                        "[SmartAPI] Login successful."
                    )

                    return {
                        "success": True
                    }

                logger.error(
                    data.get("message")
                )

                return {

                    "success": False,

                    "reason": data.get("message"),

                }

            except Exception as e:

                logger.exception(e)

                return {

                    "success": False,

                    "reason": str(e),

                }

    # --------------------------------------------------------------

    async def ensure_authenticated(
        self,
    ) -> dict[str, Any]:

        if not self.is_token_valid():

            return await self.login()

        return {
            "success": True
        }

    # --------------------------------------------------------------

    async def get_quote(

        self,

        instruments: list[dict[str, str]],

        mode: str = "FULL",

    ) -> Optional[dict]:

        auth = await self.ensure_authenticated()

        if not auth["success"]:

            return None

        try:

            async with httpx.AsyncClient(
                timeout=8
            ) as client:

                response = await client.post(

                    f"{_ANGEL_BASE}/rest/secure/angelbroking/market/v1/quote/",

                    json={

                        "mode": mode,

                        "exchangeTokens": self._group_by_exchange(
                            instruments
                        ),

                    },

                    headers=self._base_headers(
                        with_auth=True
                    ),

                )

            data = response.json()
            import json

            logger.info(
                "[SmartAPI Quote Response]\n%s",
                json.dumps(data, indent=2)
            )
            return data.get("data")

        except Exception as e:

            logger.exception(e)

            return None
    
    # --------------------------------------------------------------
    # LTP API
    # --------------------------------------------------------------

    async def get_ltp_data(
        self,
        exchange: str,
        tradingsymbol: str,
        symboltoken: str,
    ) -> Optional[dict[str, Any]]:

        auth = await self.ensure_authenticated()

        if not auth["success"]:
            return None

        try:

            async with httpx.AsyncClient(timeout=8.0) as client:

                response = await client.post(

                    f"{_ANGEL_BASE}/rest/secure/angelbroking/order/v1/getLtpData",

                    json={

                        "exchange": exchange,

                        "tradingsymbol": tradingsymbol,

                        "symboltoken": symboltoken,

                    },

                    headers=self._base_headers(
                        with_auth=True
                    ),

                )

            data = response.json()

            return data.get("data")

        except Exception as e:

            logger.exception(
                "[SmartAPI] LTP Error : %s",
                e,
            )

            return None

    # --------------------------------------------------------------
    # Option Chain API
    # --------------------------------------------------------------

    async def get_option_chain(
        self,
        symbol: str,
        expiry_date: str,
        strike_price: float,
    ) -> Optional[dict[str, Any]]:

        auth = await self.ensure_authenticated()

        if not auth["success"]:
            return None

        try:

            async with httpx.AsyncClient(timeout=10.0) as client:

                response = await client.post(

                    f"{_ANGEL_BASE}/rest/secure/angelbroking/derivatives/v1/optionchain",

                    json={

                        "name": symbol,

                        "expirydate": expiry_date,

                        "strikePrice": strike_price,

                        "optionType": "CE PE",

                    },

                    headers=self._base_headers(
                        with_auth=True
                    ),

                )

            data = response.json()

            return data.get("data")

        except Exception as e:

            logger.exception(
                "[SmartAPI] OptionChain Error : %s",
                e,
            )

            return None

    # --------------------------------------------------------------
    # Historical Candle API
    # --------------------------------------------------------------

    async def get_historical_data(
        self,
        exchange: str,
        symboltoken: str,
        interval: str,
        from_date: str,
        to_date: str,
    ) -> Optional[list[dict[str, Any]]]:

        auth = await self.ensure_authenticated()

        if not auth["success"]:
            return None

        payload = {

            "exchange": exchange,

            "symboltoken": symboltoken,

            "interval": interval,

            "fromdate": from_date,

            "todate": to_date,

        }

        try:

            async with httpx.AsyncClient(timeout=30.0) as client:

                logger.info("[Historical] Payload: %s", payload)
                headers = self._base_headers(with_auth=True).copy()

                if "Authorization" in headers:
                    headers["Authorization"] = "Bearer ****"

                logger.info(
                    "[Historical] Headers: %s",
                    headers,
                )

                response = await client.post(
                    _HISTORICAL_API,
                    json=payload,
                    headers=self._base_headers(with_auth=True),
                )

                logger.error("[Historical] Status: %s", response.status_code)
                logger.error("[Historical] Body: %s", response.text)

                if response.status_code != 200:
                    return None

                try:
                    data = response.json()
                except Exception:
                    logger.exception("Invalid JSON received")
                    return None
            if not data.get("status"):

                logger.warning(

                    "[Historical] %s",

                    data.get("message"),

                )

                return None

            candles = []

            for row in data.get("data", []):

                candles.append(
                    {
                        "datetime": row[0],
                        "open": float(row[1]),
                        "high": float(row[2]),
                        "low": float(row[3]),
                        "close": float(row[4]),
                        "volume": int(row[5]),
                        "oi": int(row[6]) if len(row) > 6 else 0,
                    }
                )

            logger.info(

                "[Historical] %d candles downloaded",

                len(candles),

            )

            return candles

        except Exception as e:

            logger.exception(

                "[Historical] %s",

                e,

            )

            return None

# ------------------------------------------------------------------
# Module Singleton
# ------------------------------------------------------------------

_client: Optional[SmartAPIClient] = None


def init_client(settings: Settings) -> SmartAPIClient:

    global _client

    _client = SmartAPIClient(settings)

    return _client


def get_client() -> SmartAPIClient:

    if _client is None:

        raise RuntimeError(
            "SmartAPIClient not initialised."
        )

    return _client
